### PR TITLE
[#251] regex lib test

### PR DESCRIPTION
**Pull request by TaimoorRana**
1. Added demand and fact cluster to CJL server
2. Downloads cluster folders
3. Takes each regex and matches it to the correct cluster file
4. Save these matches in a binary file to be loaded by the regex_lib_test
5. Regex_lib_test checks if the regexes do indeed match the corresponding cluster file.

### CODE REVIEW

File: src/ml_service/feature_extraction/post_processing/regex/regex_lib.py [hunk 1]
```diff
@@ -881,10 +881,78 @@ def sentence_finder(self, regex_name, nb_of_files):
             file.close()
         return sentences_matched
 
+    def cluster_file_finder(self, regex_name, min_match_percentage, file_path):
```
Comment (arekmano) [inline src/ml_service/feature_extraction/post_processing/regex/regex_lib.py]: IMO these methods belong in a seperate file. keep the regex_lib as small as possible; responsible only for regexes

File: src/ml_service/feature_extraction/post_processing/regex/regex_lib.py [hunk 2]
```diff
@@ -899,3 +967,11 @@ def run():
     reg_dict['MONEY_REGEX'] = regexes.MONEY_REGEX
     save = Save()
     save.save_binary('regexes.bin', reg_dict)
+
+    # Create and save regex-cluster dict
+    RegexLib().unpack_fact_demand_bin()
```
Comment (arekmano) [inline src/ml_service/feature_extraction/post_processing/regex/regex_lib.py]: same with this

File: src/ml_service/feature_extraction/post_processing/regex/regex_lib.py (hunk 1 above)
Comment (TaimoorRana) [reply src/ml_service/feature_extraction/post_processing/regex/regex_lib.py]: Good point. Will move them

File: src/ml_service/feature_extraction/post_processing/regex/regex_lib_helper.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+def cluster_regex_mapper(folder_name, min_match_percentage, nb_of_files=-1):
+    """
+    This function searches through a given folder_name in order to find all regex-cluster pair and store them in a dict
+    :param folder_name: cluster folder to search in (fact or demand)
+    :param min_match_percentage: min percentage of sentence in a cluster file that needs to match a regex
+    :param nb_of_files: number of files to search through in the folder
+    :return: dict of regex-cluster file match
+    """
+    from util.file import Path
```
Comment (arekmano) [inline src/ml_service/feature_extraction/post_processing/regex/regex_lib_helper.py]: imports at the top

File: src/ml_service/feature_extraction/post_processing/regex/regex_lib_helper.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                cluster_regex_dict[regex[0]] = [file_name]
+    return cluster_regex_dict
+
+
+def unpack_fact_demand_bin():
+    """
+    unpacks fact and demand binaries and move them to their appropriate folders
+    :return: None
+    """
+    from util.file import Path
```
Comment (arekmano) [inline src/ml_service/feature_extraction/post_processing/regex/regex_lib_helper.py]: imports at the top

**Comment by TaimoorRana** [comment]
This is the regex to cluster file mapping and match percentage.

Format: regex Name, cluster file name, percentage matched

regex name: apartment_impropre | file name matched: 900.txt | match percentage: 100.0%
regex name: asker_is_tenant | file name matched: 963.txt | match percentage: 60.0%
regex name: tenant_bad_payment_habits | file name matched: 967.txt | match percentage: 100.0%
regex name: asker_is_landlord | file name matched: 983.txt | match percentage: 60.0%
regex name: landlord_retakes_apartment | file name matched: 952.txt | match percentage: 100.0%
regex name: tenant_owes_rent | file name matched: 3485.txt | match percentage: 100.0%
regex name: tenant_landlord_agreement | file name matched: 1195.txt | match percentage: 80.0%
regex name: lack_of_proof | file name matched: 992.txt | match percentage: 75.0%
regex name: tenant_violence | file name matched: 1396.txt | match percentage: 80.0%
regex name: violent | file name matched: 1396.txt | match percentage: 80.0%
regex name: landlord_rent_change | file name matched: 2320.txt | match percentage: 100.0%
regex name: tenant_withold_rent_without_permission | file name matched: 1609.txt | match percentage: 100.0%
regex name: tenant_monthly_payment | file name matched: 972.txt | match percentage: 100.0%
regex name: tenant_rent_not_paid_more_3_weeks | file name matched: 3118.txt | match percentage: 81.81%
regex name: tenant_left_without_paying | file name matched: 2208.txt | match percentage: 92.30%
regex name: landlord_notifies_tenant_retake_apartment | file name matched: 2747.txt | match percentage: 100.0%
regex name: case_fee_reimbursement | file name matched: 2955.txt | match percentage: 86.11%
regex name: tenant_refuses_retake_apartment | file name matched: 2121.txt | match percentage: 85.71%
regex name: landlord_prejudice_justified | file name matched: 2313.txt | match percentage: 100.0%
regex name: lease | file name matched: 3323.txt | match percentage: 100.0%
regex name: tenant_declare_insalubre | file name matched: 857.txt | match percentage: 66.66%
regex name: tenant_continuous_late_payment | file name matched: 2718.txt | match percentage: 100.0%
regex name: tenant_landlord_agreement | file name matched: 90.txt | match percentage: 100.0%
regex name: tenant_bad_payment_habits | file name matched: 96.txt | match percentage: 100.0%
regex name: tenant_monthly_payment | file name matched: 341.txt | match percentage: 100.0%
regex name: tenant_rent_not_paid_more_3_weeks | file name matched: 147.txt | match percentage: 100.0%
regex name: tenant_owes_rent | file name matched: 185.txt | match percentage: 100.0%
regex name: landlord_retakes_apartment | file name matched: 74.txt | match percentage: 100.0%
regex name: asker_is_landlord | file name matched: 298.txt | match percentage: 100.0%
regex name: landlord_notifies_tenant_retake_apartment | file name matched: 331.txt | match percentage: 100.0%
regex name: lease | file name matched: 341.txt | match percentage: 80.0%
regex name: tenant_left_without_paying | file name matched: 416.txt | match percentage: 75.0%
regex name: apartment_impropre | file name matched: 426.txt | match percentage: 100.0%

**Review by arekmano — APPROVED** [review]
LGTM

**Pull request merged**